src/Marten/Events/Daemon/Internals/SubscriptionAgent.cs

```diff
@@ -0,0 +1,298 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+using JasperFx.Core;
+using Marten.Events.Projections;
+using Microsoft.Extensions.Logging;
+
+namespace Marten.Events.Daemon.Internals;
+
+public class SubscriptionAgent: ISubscriptionAgent, IAsyncDisposable
+{
+    private readonly AsyncOptions _options;
+    private readonly IEventLoader _loader;
+    private readonly ISubscriptionExecution _execution;
+    private readonly ShardStateTracker _tracker;
+    private readonly ILogger _logger;
+    public ShardName Name { get; }
+    private readonly CancellationTokenSource _cancellation = new();
+    private readonly ActionBlock<Command> _commandBlock;
+    private ErrorHandlingOptions _errorOptions = new();
+    private IDaemonRuntime _runtime = new NulloDaemonRuntime();
+
+    public SubscriptionAgent(ShardName name, AsyncOptions options, IEventLoader loader,
+        ISubscriptionExecution execution, ShardStateTracker tracker, ILogger logger)
+    {
+        _options = options;
+        _loader = loader;
+        _execution = execution;
+        _tracker = tracker;
+        _logger = logger;
+        Name = name;
+
+        _commandBlock = new ActionBlock<Command>(Apply, _cancellation.Token.SequentialOptions());
+
+        ProjectionShardIdentity = name.Identity;
+        if (_execution.DatabaseName != "Marten")
+        {
+            ProjectionShardIdentity += $"@{_execution.DatabaseName}";
+        }
+    }
+
+    public string ProjectionShardIdentity { get; private set; }
+
+    public CancellationToken CancellationToken => _cancellation.Token;
+
+    // Making the setter internal so the test harness can override it
+    // It's naughty, will make some people get very upset, and
+    // makes unit testing much simpler. I'm not ashamed
+    public long LastEnqueued { get; internal set; }
+
+    public long LastCommitted { get; internal set; }
+
+    public long HighWaterMark { get; internal set; }
+
+    public async Task ReportCriticalFailureAsync(Exception ex)
+    {
+        try
+        {
+#if NET8_0_OR_GREATER
+            await _cancellation.CancelAsync().ConfigureAwait(false);
+#else
+            _cancellation.Cancel();
+#endif
+            await _execution.HardStopAsync().ConfigureAwait(false);
+            PausedTime = DateTimeOffset.UtcNow;
+            Status = AgentStatus.Paused;
```

Comment:
**QUESTION:** What does the paused status mean in this context? Will it also be used in the future for the pausing and stopping agent? Currently, I see that's only paused when there was an error in the processing. Then it'll be restarted in ProjectionCoordinator, but if the issue is non-transient, then it will always fail again. 

Am I missing something? Maybe it's worth having differentiation between paused and errored?

Reply:
"Paused" absolutely just means "temporarily stopped because of an error" here.

I think I'd like to talk this through with you over Zoom. This and https://github.com/JasperFx/marten/issues/2952 and what monitoring might be for this stuff. Also how it's going to be different in "Critter Stack Pro".

Let's track this work here: https://github.com/JasperFx/marten/issues/2968

Reply:
Ok, I'm curious how that will work. I'm fine with moving that to other dedicated issues.